functions/go/set-image/transformer/images_transformer.go

```diff
@@ -0,0 +1,170 @@
+package transformer
+
+import (
+	"fmt"
+
+	"github.com/GoogleContainerTools/kpt-functions-catalog/functions/go/set-image/custom"
+	"github.com/GoogleContainerTools/kpt-functions-catalog/functions/go/set-image/image_util"
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+)
+
+// SetImage supports the set-image workflow, it uses Config to parse functionConfig, Transform to change the image
+type SetImage struct {
```

Comment:
(I think we've talked about the `SetImage` attributes structure and the long term plans for `additionalImageField` as well?)  
The general ideas are:
1. The core function configs like `newImage` `containerName` `imageName` should live in this file. The main focus of the setImage is for core k8s resources.
2. We keep the `additionalImageField` for backward compatibility reasons. Eventually it will be dropped once we have a solution for CRD resources like supporint OpenAPI schema.

So the code should be structured in a way that can clearly distinguish the `additionalImageFields` and the core k8s resource logics, and dropping the `additionalImageField` interface should not require any code refactoring, or even changes except removing the `additionalImageField` line from `SetImage` struct

Reply:
Based on that, the code should be look like 
```go
// image_transformer.go
type SetImage struct {
   Image Image `json:"image,omitempty" yaml:"image,omitempty"`

   ... 
   AdditionalImageFields types.FsSlice `json:"additionalImageFields,omitempty" yaml:"additionalImageFields,omitempty"`
}

type Image struct {
    Name string `json:"name,omitempty" yaml:"name,omitempty"`
    NewName string `json:"newName,omitempty" yaml:"newName,omitempty"`
    NewTag string `json:"newTag,omitempty" yaml:"newTag,omitempty"`
    Digest string `json:"digest,omitempty" yaml:"digest,omitempty"`
}

func (r *SetImage) Run(...) {
    if t.AdditionalImageFields != nil {
        custom.SetAdditionalFieldSpec(t.GetMap("image"), items, t.AdditionalImageFields, t.context)
    }
}
```
// custom.go
func SetAdditionalFieldSpec(imageObject fn.SubObject, objects fn.KubeObjects, addImgFields types.FsSlice, ctx *fn.Context) {
    image, additionalImageFields := transformStruct(img, addImgFields)

    for i, obj := range objects {
        objRN, err := yaml.Parse(obj.String())
        if err != nil {
            ctx.ResultErr(err.Error(), obj)
        }
        filter := imagetag.Filter{
            ImageTag: types.Image{
                Name:    imageObject.NestedStringOrDie("name"),
                NewName: imageObject.NestedStringOrDie("newName"),
                NewTag:  imageObject.NestedStringOrDie("newTag"),
                Digest:  imageObject.NestedStringOrDie("digest"),
            },
            FsSlice:  additionalImageFields,
        }

        filter.WithMutationTracker(LogResultCallback(ctx, obj))
        err = filtersutil.ApplyToJSON(filter, objRN)
        if err != nil {
            ctx.ResultErr(err.Error(), obj)
        }
        newObj, err := fn.ParseKubeObject([]byte(objRN.MustString()))
        if err != nil {
            ctx.ResultErr(err.Error(), obj)
        }
        objects[i] = newObj
    }
}
```

Reply:
This structure is pretty much what I have here. Except I have the image definition inside a separate package. Because `custom` package need to import `Image` struct, but `transform` package already imports `custom`, to avoid dependency error I have a separate package `image_util` and put `Image` struct here, so that both `transformer` and `custom` can import

Reply:
there are two major differences (See my first comment about the general ideas). 
1. The dependency error should not be the reason to duplicate code from another library. Suggest taking a deeper look at the parameters I used in the code, I use `t.GetMap("image")` and regenerate the `types.Image` via fn.SubObject, so it won't trigger dependency loop.
2. "dropping the additionalImageField interface should not require any code refactoring, or even changes". In this PR, the code needs a refactoring because importing "image_utiles" is no longer a necessary workaround for depedency loop but split two closely related structure `SetImage` and `Image` apart.

Reply:
I see your point! One other thing that I would like to ask about. I see you use `types.FsSlice` in `transformer`, but it is best to avoid kustomize pacakges here. Can I have my duplicate definition of `FsSlice` inside `image_util` here?

Reply:
`t.GetMap("image")` --> `fnConfig.GetMap("image")` The fnConfig is the second parameter in the `Run` method.

If the tradeoff is between introducing `image_util` and  defining `types.FnSlice` in `SetImage` struct, I would prefer we choose the latter. For `image_util` the right way is to fork kustomize like @natasha41575 suggested, not borrow the code directly. I don't see value to introduce this complexity TBH.

Reply:
I apply the same way for `FsSlice`, using `fn.SliceSubObjects`